### ISSUE TITLE
Obsolete WebHostBuilder

### ISSUE THREAD

**Issue by Tratcher** [dotnet/aspnetcore#20964]
WebHostBuilder was replaced by HostBuilder in 3.0. In 5.0 we should consider obsoleting the old one.

**Comment by anurse** [comment]
To clarify even more, the idea here is to obsolete the types and public APIs on:

* [WebHost](https://github.com/dotnet/aspnetcore/blob/master/src/DefaultBuilder/src/WebHost.cs)
* [WebHostBuilder](https://github.com/dotnet/aspnetcore/blob/master/src/Hosting/Hosting/src/WebHostBuilder.cs)
* Add an "obsolete" banner to https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/web-host?view=aspnetcore-3.1

The obsolete message should include an `aka.ms` link that points to https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-3.1#settings-for-web-apps

**Comment by davidfowl** [comment]
We shouldn’t do this until we resolve the remaining gaps that people complain about:
- Passing extra state to the startup class (for e.g)

**Comment by Tratcher** [comment]
> * Passing extra state to the startup class

Related issue: https://github.com/dotnet/aspnetcore/issues/19809

**Comment by davidfowl** [comment]
Also the `IStartup` replacement.

**Comment by BrennanConroy** [comment]
Lets start cleaning up areas before obsoleting this starting in preview6.

- [x] SignalR - https://github.com/dotnet/aspnetcore/pull/22602
- [x] DataProtection #24254
- [x] Middleware - https://github.com/dotnet/aspnetcore/pull/23626
  - [x] Rewrite
  - [x] Localization
  - [x] Localization.Routing
  - [x] StaticFiles
  - [x] HttpsPolicy
  - [x] HttpOverrides
  - [x] WebSockets
  - [x] Diagnostics
  - [x] HostFiltering
  - [x] ResponseCaching
  - [x] Session
  - [x] HealthChecks
  - [x] CORS
  - [x] Diagnostics.EntityFrameworkCore
  - [x] MiddlewareAnalysis
  - [x] ResponseCompression
  - [x] ConcurrencyLimiter
  - [x] NodeServices
  - [x] HeaderPropagation
- [x] Kestrel #24279
- [x] IIS https://github.com/dotnet/aspnetcore/pull/24280
- [x] Routing #24281
- [x] Security
  - [x] Authentication https://github.com/dotnet/aspnetcore/pull/23783
  - [x] CookiePolicy https://github.com/dotnet/aspnetcore/pull/24282
- [x] Identity https://github.com/dotnet/aspnetcore/pull/23783
- [x] Azure https://github.com/dotnet/aspnetcore/pull/24283
- [x] MVC https://github.com/dotnet/aspnetcore/pull/62703
- [x] MusicStore #24284
- [x] HttpSys #24285
- [x] Hosting #24297

**Comment by BrennanConroy** [comment]
Example changes https://github.com/dotnet/aspnetcore/pull/23626

Make sure to dispose the IHost.

Common changes:

Before:
```c#
var host = new WebHostBuilder()
    .Configure(app =>
    {
        //...
    }).Build();
```
After:
```c#
using var host = new HostBuilder()
    .ConfigureWebHost(webHostBuilder =>
    {
        webHostBuilder
        .Configure(app =>
        {
            //...
        }
    }).Build();
```

TestServer:
Before:
```c#
var builder = new WebHostBuilder();
var server = new TestServer(builder);
```
After:
```c#
using var host = new HostBuilder()
    // ...
    .UseTestServer()
    .Build();
var server = host.GetTestServer();
```

**Comment by BrennanConroy** [comment]
@mkArtakMSFT Can you find some folks to handle MVC?
@HaoK Can you take care of Identity/Security?

**Comment by HaoK** [comment]
Yup I can do identity/security

**Comment by davidfowl** [comment]
Just to be clear, we're not obsoleting anything in 5.0 😄. We haven't scheduled the linked items though I am supportive of updating the other projects.

**Comment by HaoK** [comment]
How bad is it if the host doesn't get disposed?  There's a bunch of tests that rely on a helper method that creates the server, which if the helper has a using var host, the host gets disposed and the test code fails due to object disposed.

i.e. 
```
   Task<TestServer> CreateServer() { 
      using var host = ... .Build(); 
      await host.StartAsync();
      return host.GetTestServer();
```

**Comment by BrennanConroy** [comment]
You'll get hangs because by default the console lifetime service is hooked up which waits for ctrl-c (or dispose) before closing.

I changed a bunch of helpers in middleware to return `IHost` instead of `TestServer`.

**Comment by HaoK** [comment]
Hangs?  Everything seems to work fine if I just drop the using

**Comment by HaoK** [comment]
meaning the tests complete just fine without the using

**Comment by BrennanConroy** [comment]
Yeah, I'm not sure how VS handles that, but when running from command line you should see hangs.

**Comment by HaoK** [comment]
I just tried via dotnet test and it 'worked'.  Does that mean its ok?

**Comment by Tratcher** [comment]
@BrennanConroy other thing we should consider is having UseTestServer register its own lifetime to replace ConsoleLifetime. The implementation should be a no-op.

**Comment by BrennanConroy** [comment]
So, now there isn't risk of hangs for not disposing since we just merged the `NoopHostLifetime`. However, I'd still suggest disposing it if it's not too hard.

**Comment by BrennanConroy** [comment]
> @mkArtakMSFT Can you find some folks to handle MVC?

Ping

**Comment by Kahbazi** [comment]
Is anybody working on MVC?

**Comment by Tratcher** [comment]
> Is anybody working on MVC?

Not yet. MVC is a very large area with some complicated tests. I'd suggest splitting it into multiple PRs by project type (e.g. samples, tests, etc.) or feature area. Also, when you find something more unusual like WebApplicationFactory, only fix one test and submit a draft PR to make sure that's how they want it.

**Comment by BrennanConroy** [comment]
* Soft obsolete (don't delete)
* Analyzer?
* Code fixes

**Comment by joperezr** [comment]
Closing as this one is done and announcement is live now.
